### ISSUE TITLE
Synchronize audio streams based on their frequencies

### ISSUE THREAD

**Issue by marioortizmanero** [marioortizmanero/spotify-music-videos#8]
Issue to keep track of the progress on audio synchronziation, inspired by this [project](https://github.com/allisonnicoledeal/VideoSync) and thanks to pinkman on the AUR comments, who linked it.

Spotify has an audio analysis feature that could give [more information about the song](https://developer.spotify.com/documentation/web-api/reference/tracks/get-audio-analysis/) It can provide us with [rythm, pitch and timbre information](https://developer.spotify.com/documentation/web-api/reference/tracks/get-audio-analysis/#rhythm).

But if we rely on the Spotify API then we won't be able to implement this for Linux (DBus API) users.

And the YouTube video is played with VLC directly, so the videos aren't fully downloaded and it may make the audio analysis harder to do. We would need a more dynamic way to do it on the go. That, or change how the videos are played.

I think the bigger issue right now is the latter. How can we analyze a video that isn't even downloaded yet? 

EDIT: This is being worked on at https://github.com/marioortizmanero/vidify-audiosync

**Comment by akaessens** [comment]
Hi,

First Problem:
The spotify audio could be be recorded directly from the audio sink of the operating system and fed into the audio signal processing.

As for the second one, i was thinking about a virtual audio loopback device but that wolud require some administrative privileges which is not an adequate solution. Maybe it would be enough to download a few seconds of the youtube audio but that will probably fail for a significant amount of music videos because of intros etc.

**Comment by marioortizmanero** [comment]
1. Recording the Spotify audio looks like the simplest way to support it on Linux but it may consume many resources. Some benchmarks will have to be done to determine how much it actually affects the performance because some users may prefer to activate this manually if it's a big difference.

2. Downloading part of the video doesn't really sound like a bad idea to me. Most of these cinematic intros don't last more than 15-20 seconds. I would actually start with this idea to see how reliable the synchronization would be before implementing something more complex.

Another way of doing it is buffering the video inside the program instead of having VLC do it, so that the data can be accessed directly instead of having to do workarounds to get the VLC audio. But I have no idea of how that could be done and it may bring some bugs.

**Comment by marioortizmanero** [comment]
Very simple overview using the Spotify API to check how precise it is. Here's an example of the analysis with [Des Rocks - Let Me Live / Let me Die](https://www.youtube.com/watch?v=Ozo8LyvS4fE): 

Raw data:
https://0x0.st/zW2U.json

**Plots with the first 30 seconds of segments and with a confidence greater than 0.6:**

Max volume (segments > loudness_max):
![maxvol](https://i.imgur.com/TaIwlfI.png)

Starting volume (segments > loudness_start):
![maxvol](https://i.imgur.com/ZfRBKOG.png)

Audacity waveform (dB) from youtube:
![audacity](https://i.imgur.com/kGurbjt.png)

So yeah loudness doesn't look like a reliable way to sync. The starting volume seems to be more precise though.

Timbre (segments > timbre[i]):
![timbre](https://i.imgur.com/1sx57FM.png)

Pitch (segments > pitches[i]):
![pitch](https://i.imgur.com/zUimQZn.png)

The pitches are a mess, but the timbre could be useful if we manage to get the timbres from the youtube video too... Using the Spotify API seems really complicated though.

**Comment by akaessens** [comment]
> After a quick benchmark, recording on background doesn't seem to consume as many resources as I thought. Something like `ffmpeg -f pulse -i default -t 15 spotify.wav` works but it would have to be made multi-platform.

[multiplatform library for recording audio](https://github.com/bastibe/SoundCard)

**Comment by marioortizmanero** [comment]
After 2.0 I'm going to continue this. For now I'm looking for a good library to record the audio, because @akaessens, the one you linked can't record the speaker, only the microphone.

The most used option is [pyaudio](https://github.com/jleb/pyaudio) but it requires [portaudio](http://www.portaudio.com/). Maybe there's something similar that doesn't.

[python-sounddevice](https://python-sounddevice.readthedocs.io/en/latest/) is also an option. It also requires portaudio but it seems easier to use.

Either that, or just using ffmpeg.

**Comment by marioortizmanero** [comment]
Update!!!

I've managed to implement the FFT itself already. I didn't expect it to be that easy and that precise! It needs lots of polishing, and the align part isn't finished yet, but the outputs are promising:

![plot](https://i.imgur.com/pjd8hIH.png)

The visualizations on the left are from Audacity, the program's are on the right. The top track was downloaded from YouTube and the second was recorded from the desktop.

Also, it's pretty fast! It concurrently runs the audio processing for both files. The FFT is done with FFTW, the "Fastest Fourier Transformation in the West" (can confirm it's true).

TODO list:
 - [ ] Find a better WAV reading library than `AudioFile.h` or fix it to support multi-threading.
 - [ ] Align the tracks
 - [ ] Record the desktop audio (if possible, this should be done with a C++ library to have access to the raw data instead of calling `ffmpeg` and then reading the file)
 - [ ] Play the videos from a buffer instead of with VLC/mpv (generated inside the C++ program, since I later need its raw data directly inside the program)
 - [ ] How much should be recorded before starting the audio sync program. Is 15 seconds enough?
 - [ ] Turn it into a C++ module
 - [ ] Integrate it in the app
 - [ ] Tests

**Comment by akaessens** [comment]
Hello Mario, this is some promising progress.

About the FFT size: 1024-point FFT is common in audio processing and with a sampling rate of 48 kHz it gives ~21 ms long "slices" of the audio file which are calculated.

How do you think about the following problem: 
For the first 15 seconds of the spotify track (for the recording) there will be no sync possibility, should the video be shown unsynced for the start? 

Instead of multithreading on a long sequence i suggest something like the following
- spotify video starts
- start downloading
- in parallel three threads:
1. record a short snippet (maybe 3 s) from the spotify playback
2. continuously calculate fft of the downloaded video and save results
3. fft of the short snippet and try matching with the already known results 

There probably must be some confidence threshold and silence at start must be cut off, too.


When you go the way of multiprocessing on the GPU, maybe C++ with OpenGL might be worth a look (Quick search results in this: [GLFFT](https://github.com/Themaister/GLFFT))

**Comment by marioortizmanero** [comment]
The bottleneck here is that the desktop audio is being recorded, and I can't just speed that up. That's why I also tried using the Web API in the beginning. Your idea is much more viable after seeing how fast and efficient the algorithm can be, so it could definitely be a good plan.

I currently calculate the FFT with FFTW. But seems like using FFTW on a GPU is [not doable](https://stackoverflow.com/questions/16780258/running-fftw-on-gpu-vs-using-cufft); something like CUFFT has to be used instead. But both CUFFT and GLFFT only have support for Nvidia. And I don't want the GPU method to only be available for one system. Besides, I have an AMD GPU so I wouldn't even be able to test it myself. I'll have to look for another way to do it.

First of all I have to do the audio matching, though. Which seems like an ideal part where AI could be used. But I don't know much about AI either so I'll use a regular algorithm for now. There's probably something similar online I can base on anyway.

In the end, I haven't strictly followed [Allison's project](https://github.com/allisonnicoledeal/VideoSync) because I don't really get why she splits the audio both vertically and horizontally to do the matching. Right now I'm only splitting the audio horizontally (in time intervals) and normalize the data in respect to the maximum frequency found.

**Comment by akaessens** [comment]
I guess this is your [Cross-correlation Source](https://dsp.stackexchange.com/questions/736/how-do-i-implement-cross-correlation-to-prove-two-audio-files-are-similar). This definitely is the best way of calculating the offset, also much better than any neural network in my opinion.

It should work without FFT, too.
- numpy.correlate(a,b,"full") 

The output vector must be interpreted as "how similar are the arrays when they are shifted by n positions. The index of the maximum is the delay in sample points.

Could you send me your sample .wav pair?

**Comment by marioortizmanero** [comment]
I've tried to implement numpy's correlate in the C++ script but it's been quite painful. For some reason the confidence results aren't between 0 and 1 so they must be wrong. I don't even know if it's going to be any faster anyway. I'll investigate about numpy's implementation.

Allison's project was more confusing than helpful to be honest. It took me way too long to find out about Cross-Correlation, too.

Here you have the WAV files I'm using:

https://drive.google.com/drive/folders/1fnq60CGV0m8BtHXVlHoU6ArKXkdbf-HX?usp=sharing

**Comment by akaessens** [comment]
Yes i just had a similar script and have the same issues.

A useful information can be gathered from the position where the corr array first is bigger that maybe 20% of the maximum, this is where the signals start to become similar -> This is the zero-offset (but that could just be cut off anyways).

I will be looking into that further

**Comment by marioortizmanero** [comment]
I could also keep going with the C++ extension with FFT. It. It's not even 100% concurrent yet and it's orders of magnitude faster. I've made the code public in [marioortizmanero/audiosync](https://github.com/marioortizmanero/audiosync), but it's nowhere near being finished and there's lots of debug messages everywhere. Just so that you can take a look if you want.

For now I'm trying to wrap my mind around the [Circular convolution theorem and cross-correlation theorem](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Circular_convolution_theorem_and_cross-correlation_theorem) because it seems I don't know enough math yet...

**Comment by marioortizmanero** [comment]
Update: I have fixed the algorithm (only the python sketch for now), it was incorrect. It now works PERFECTLY. It's amazing how good the results are:

![img](https://i.imgur.com/gXgJ7SF.png)

And that's only with 5 seconds! I also learned more about the Pearson & Spearman coefficients to obtain a confidence of how good the results are. Now I have to figure out a couple more things:

* If it's worth it using C: the results are actually really fast with numpy and scipy. I'm not too experienced with C so It might even be slower.

* If I should use the Pearson or the Spearman coefficient

* Recording the audio and downloading the video. Both have to be the same format, too.

**Comment by marioortizmanero** [comment]
Support for audio synchronization has been added to the [2.0 release](https://github.com/marioortizmanero/spotify-music-videos/pull/25). I documented it as a Work In Progress feature, because It still needs a lot of work but more or less it works. If I release it in the next version it might get more attention & feedback. I'm closing this issue, as this is being continued in its repo issues: https://github.com/marioortizmanero/vidify-audiosync
